File: src/backend/llvm.rs
use std::collections::HashMap;
use std::ffi::{CStr, CString};
use std::os::raw::{c_char, c_uint};
use std::path::Path;
use std::sync::Arc;

use llvm_sys::*;
use llvm_sys::core::*;
use llvm_sys::error::*;
use llvm_sys::prelude::*;
use llvm_sys::target::*;
use llvm_sys::target_machine::*;

use crate::ir::*;
use crate::ir::types::*;
use std::ops::Deref;
use std::ptr;

pub struct LLVMBackend<'compiler> {
    compiler: &'compiler Compiler,
    module: &'compiler Module,
    context: LLVMContextRef,
    builder: LLVMBuilderRef,
    llvm_module: LLVMModuleRef,
    type_cache: HashMap<Type, LLVMTypeRef>,
    dec_cache: HashMap<DeclarationId, (LLVMValueRef, LLVMTypeRef)>,
}

impl<'c> LLVMBackend<'c> {
    pub fn new(compiler: &'c Compiler, module: &'c Module) -> Self {
        unsafe {
            LLVMInitializeX86TargetInfo();
            LLVMInitializeX86Target();
            LLVMInitializeX86TargetMC();
            LLVMInitializeX86AsmPrinter();
            let context = LLVMContextCreate();
            let module_name = CString::new(module.name.clone()).unwrap();
            let llvm_module: LLVMModuleRef = LLVMModuleCreateWithNameInContext(module_name.as_ptr(), context);
            let builder: LLVMBuilderRef = LLVMCreateBuilder();

            /*
            let target_triple = LLVMGetDefaultTargetTriple();
            let mut target = ptr::null_mut();
            let mut error = ptr::null_mut();
            LLVMGetTargetFromTriple(target_triple, target, error);
            LLVMDisposeErrorMessage(*error);
            let target_machine = LLVMCreateTargetMachine(*target, target_triple, cstr(""), cstr(""), LLVMCodeGenOptLevel::LLVMCodeGenLevelNone, LLVMRelocMode::LLVMRelocDefault, LLVMCodeModel::LLVMCodeModelDefault);
            LLVMDisposeMessage(target_triple);
            */

            Self {
                compiler,
                module,
                context,
                builder,
                llvm_module,
                type_cache: HashMap::new(),
                dec_cache: HashMap::new(),
            }
        }
    }

    unsafe fn convert_type(&mut self, typ: Type) -> LLVMTypeRef {
        if let Some(typ) = self.type_cache.get(&typ) {
            return *typ;
        }
        let llvm_type: LLVMTypeRef = match typ.clone() {
            Type::Primitive(p) => {
                match p {
                    PrimitiveType::Int(b) => {
                        LLVMIntTypeInContext(self.context, b as u32)
                    }
                    PrimitiveType::Float(b) => {
                        match b {
                            16 => LLVMHalfTypeInContext(self.context),
                            32 => LLVMFloatTypeInContext(self.context),
                            64 => LLVMDoubleTypeInContext(self.context),
                            80 => LLVMX86FP80TypeInContext(self.context), // TODO: research this
                            128 => LLVMFP128TypeInContext(self.context),
                            b => panic!(format!("couldn't convert float type {:?}", b)),
                        }
                    }
                    PrimitiveType::Bool => {
                        LLVMIntTypeInContext(self.context, 1)
                    }
                    p => panic!(format!("couldn't convert primitive type {:?}", p))
                }
            }
            Type::Function(f) => {
                let mut args = Vec::with_capacity(f.arguments.len());
                for arg in f.arguments.iter() {
                    args.push(self.convert_type(arg.deref().clone()));
                }
                let result_type = self.convert_type(f.result.deref().clone());
                LLVMFunctionType(result_type, args.as_mut_ptr(), args.len() as c_uint, 0)
            }
            t => panic!(format!("couldn't convert type {:?}", t))
        };

        self.type_cache.insert(typ, llvm_type);
        llvm_type
    }

    pub fn process(&mut self) {
        // process "headers" of declarations first
        for declaration in self.module.declarations.iter() {
            self.process_declaration_header(declaration);
        }
        for declaration in self.module.declarations.iter() {
            self.process_declaration(declaration);
        }
    }

    /// Generate anything needed to refer to a declaration out of order, like
    /// functions calling later defined functions needing a declaration
    fn process_declaration_header(&mut self, declaration: &Declaration) {
        let declaration_id = DeclarationId::from(Some(self.module), declaration);
        match declaration {
            Declaration::Function(ref function) => {
                let function_name = CString::new(function.name.clone()).unwrap();
                self.process_function_header(declaration_id, function_name, function.as_ref());
            }
            // TODO: cache struct types?
            _ => {}
        }
    }

    /// Declares the function in the LLVM module without creating a body for it.
    fn process_function_header(
        &mut self,
        declaration_id: DeclarationId,
        function_name: CString,
        function: &Function,
    ) {
        unsafe {
            let function_type = function.get_type(self.compiler);
            let llvm_function_type: LLVMTypeRef = self.convert_type(function_type.deref().clone());
            let llvm_function: LLVMValueRef = LLVMAddFunction(self.llvm_module, function_name.as_ptr(), llvm_function_type);
            self.dec_cache.insert(declaration_id, (llvm_function, llvm_function_type));
        }
    }

    fn process_declaration(&mut self, declaration: &Declaration) {
        let declaration_id = DeclarationId::from(Some(self.module), declaration);
        match declaration {
            Declaration::Function(ref function) => {
                let function_name = CString::new(function.name.clone()).unwrap();
                self.process_function_body(declaration_id, function_name, function.as_ref());
            }
            _ => {}
        }
    }

    fn process_function_body(
        &mut self,
        declaration_id: DeclarationId,
        function_name: CString,
        function: &Function,
    ) {
        unsafe {
            let (llvm_function, llvm_function_type) = self.dec_cache.get(&declaration_id).unwrap().clone();
            LLVMSetFunctionCallConv(llvm_function, LLVMCallConv::LLVMCCallConv as u32);
            self.process_blocks(declaration_id, function_name, function, llvm_function, function.blocks.as_slice());
        }
    }

    fn process_blocks(
        &mut self,
        declaration_id: DeclarationId,
        function_name: CString,
        function: &Function,
        llvm_function: LLVMValueRef,
        blocks: &[Block],
    ) {
        let mut block_map: HashMap<usize, LLVMBasicBlockRef> = HashMap::new();
        let mut instruction_map: HashMap<usize, LLVMValueRef> = HashMap::new();

        for (block_id, _) in blocks.iter().enumerate() {
            let block_name = CString::new(format!("block{}", block_id)).unwrap();
            let llvm_block = unsafe { LLVMAppendBasicBlock(llvm_function, block_name.as_ptr()) };
            block_map.insert(block_id, llvm_block);
        }

        for (block_id, block) in blocks.iter().enumerate() {
            let llvm_block = *block_map.get(&block_id).unwrap();
            unsafe {
                LLVMPositionBuilderAtEnd(self.builder, llvm_block);

                for (instruction_id, instruction) in block.instructions.iter().enumerate() {
                    let instruction_id = instruction_id + block.ins_start_offset;
                    let instruction_name = CString::new(format!("ins{}", instruction_id)).unwrap();

                    match *instruction {
                        Instruction::IntegerLiteral(ref i) => {
                            let int_type = self.convert_type(Type::Primitive(PrimitiveType::Int(32)));
                            let value = LLVMConstInt(int_type, i.0 as u64, 0);
                            instruction_map.insert(instruction_id, value);
                        }
                        Instruction::BooleanLiteral(ref b) => {
                            let bool_type = self.convert_type(Type::Primitive(PrimitiveType::Bool));
                            let value = LLVMConstInt(bool_type, if b.0 { 1 } else { 0 }, 0);
                            instruction_map.insert(instruction_id, value);
                        }
                        Instruction::Alloca(ref a) => {
                            let typ = block.get_instruction(a.reference_ins).get_type(self.compiler, block);
                            let llvm_type = self.convert_type(typ.deref().clone());
                            let value = LLVMBuildAlloca(self.builder, llvm_type, instruction_name.as_ptr());
                            instruction_map.insert(instruction_id, value);
                        }
                        Instruction::Store(ref s) => {
                            let ptr = *instruction_map.get(&s.ptr.0).unwrap();
                            let val = *instruction_map.get(&s.value.0).unwrap();
                            let value = LLVMBuildStore(self.builder, val, ptr);
                            instruction_map.insert(instruction_id, value);
                        }
                        Instruction::Load(ref l) => {
                            // let typ = instruction.get_type_with_context(self.compiler, block, function);
                            // let typ = block.get_instruction(l.reference_ins).get_type(self.compiler, block);
                            // let llvm_type = self.convert_type(typ.deref().clone());
                            let ptr = *instruction_map.get(&l.ptr.0).unwrap();
                            let value = LLVMBuildLoad(self.builder, ptr, instruction_name.as_ptr());
                            instruction_map.insert(instruction_id, value);
                        }
                        Instruction::GetParameter(ref g) => {
                            for (i, (name, _)) in function.arguments.iter().enumerate() {
                                if name == &g.name {
                                    let value = LLVMGetParam(llvm_function, i as c_uint);
                                    instruction_map.insert(instruction_id, value);
                                    break;
                                }
                            }
                        }
                        Instruction::Jump(ref j) => {
                            let to_llvm_block = block_map.get(&j.block.0).unwrap();
                            LLVMBuildBr(self.builder, *to_llvm_block);
                        }
                        Instruction::Branch(ref b) => {
                            let cond = *instruction_map.get(&b.condition.0).unwrap();
                            let true_block = *block_map.get(&b.true_block.0).unwrap();
                            let false_block = *block_map.get(&b.false_block.0).unwrap();
                            LLVMBuildCondBr(self.builder, cond, true_block, false_block);
                        }
                        Instruction::Return(ref r) => {
                            let value = *instruction_map.get(&r.instruction.0)
                                .expect(&format!("couldn't find instruction: {}", &r.instruction.0));
                            LLVMBuildRet(self.builder, value);
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    pub fn dump(&self) {
        unsafe {
            LLVMDumpModule(self.llvm_module);
        }
    }

    pub fn emit<S: AsRef<str>>(&mut self, path: S) {
        unsafe {
            let error = ptr::null_mut();
            let cpath = CString::new(path.as_ref()).unwrap();
            LLVMPrintModuleToFile(self.llvm_module, cpath.as_ptr(), error);
            LLVMDisposeErrorMessage(*error);
        }
    }
}

impl Drop for LLVMBackend<'_> {
    fn drop(&mut self) {
        unsafe {
            LLVMDisposeModule(self.llvm_module);
            LLVMDisposeBuilder(self.builder);
            LLVMContextDispose(self.context);
        }
    }
}

File: src/main/kotlin/dev/alox/compiler/backend/llvm.kt
package dev.alox.compiler.backend

import dev.alox.compiler.ir.IrCompiler
import dev.alox.compiler.ir.IrModule
import org.bytedeco.javacpp.PointerPointer
import org.bytedeco.llvm.LLVM.*
import org.bytedeco.llvm.global.LLVM.*

/**
 * Translates an IR Module to LLVM IR
 */
class LLVMBackend(private val compiler: IrCompiler, private val irModule: IrModule) {

    private val context: LLVMContextRef
    private val builder: LLVMBuilderRef
    private val llvmModule: LLVMModuleRef
    private val typeCache = mutableMapOf<IrModule.Declaration.Type, LLVMTypeRef>()
    private val declarationCache = mutableMapOf<IrModule.Declaration, Pair<LLVMValueRef?, LLVMTypeRef>>()

    init {
        LLVMInitializeX86TargetInfo()
        LLVMInitializeX86Target()
        LLVMInitializeX86TargetMC()
        LLVMInitializeX86AsmPrinter()
        context = LLVMContextCreate()
        llvmModule = LLVMModuleCreateWithNameInContext(irModule.name, context)
        builder = LLVMCreateBuilderInContext(context)
    }

    fun close() {
        LLVMDisposeModule(llvmModule)
        LLVMDisposeBuilder(builder)
        LLVMContextDispose(context)
    }

    private fun convertType(type: IrModule.Declaration.Type, parent: IrModule.Declaration.Type? = null): LLVMTypeRef? {
        return when (type) {
            is IrModule.Declaration.Type.Primitive.Void -> LLVMVoidTypeInContext(context)
            is IrModule.Declaration.Type.Primitive.IntT -> LLVMIntTypeInContext(context, type.bits)
            is IrModule.Declaration.Type.Primitive.FloatT -> {
                when (type.bits) {
                    16 -> LLVMHalfTypeInContext(context)
                    32 -> LLVMFloatTypeInContext(context)
                    64 -> LLVMDoubleTypeInContext(context)
                    128 -> LLVMFP128TypeInContext(context)
                    else -> null
                }
            }
            is IrModule.Declaration.Type.Primitive.Bool -> LLVMIntTypeInContext(context, 1)
            is IrModule.Declaration.Type.Ref -> {
                getType(type.inner)?.let { LLVMPointerType(it, 0) }
            }
            is IrModule.Declaration.Type.Function -> {
                val argTypes = type.declaration.arguments
                    .mapNotNull { it.declarationRef.toType() }
                    .toMutableList()
                if (parent != null) {
                    argTypes.add(0, parent)
                }

                val llvmArgTypes = PointerPointer<LLVMTypeRef>(
                    *argTypes
                        .map { getType(it) }
                        .toTypedArray()
                )

                val returnType = type.declaration.returnType.toType()
                    ?.let { getType(it) }

                LLVMFunctionType(returnType, llvmArgTypes, argTypes.size, 0)
            }
            is IrModule.Declaration.Type.Struct -> {
                val struct = type.declaration
                val llvmStruct = LLVMStructCreateNamed(context, struct.name)

                // convert the fields to LLVMTypeRefs by resolving them using the compiler and converting them
                // TODO: support type parameters
                val fieldTypes = struct.fields.mapNotNull { it.declarationRef.toType() }

                val fields = PointerPointer<LLVMTypeRef>(
                    *fieldTypes.map { getType(it) }.toTypedArray()
                )

                LLVMStructSetBody(llvmStruct, fields, fieldTypes.size, 0)
                llvmStruct
            }
            else -> {
                println("Failed to find LLVM Type for $type")
                null
            }
        }
    }

    /**
     * Convert from our IR Type to the LLVM IR Type and use the cache
     */
    fun getType(type: IrModule.Declaration.Type, parent: IrModule.Declaration.Type? = null): LLVMTypeRef? {
        var llvmType = typeCache[type]
        if (llvmType == null) {
            llvmType = convertType(type, parent)
            if (llvmType != null) {
                typeCache[type] = llvmType
            }
        }
        return llvmType
    }

    fun getType(declarationRef: IrModule.DeclarationRef, parent: IrModule.Declaration.Type? = null): LLVMTypeRef? {
        return compiler.resolve(irModule, declarationRef)?.let {
            if (it is IrModule.Declaration.Type) {
                getType(it, parent)
            } else {
                null
            }
        }
    }

    fun process() {
        irModule.declarations.forEach {
            processDeclarationHeader(it)
        }
        irModule.declarations.forEach {
            processDeclaration(it)
        }
    }

    /**
     * Generate anything needed to refer to a declaration out of order, like
     * functions calling later defined functions needing a declaration
     */
    private fun processDeclarationHeader(
        declaration: IrModule.Declaration,
        parentStruct: IrModule.Declaration.Struct? = null
    ) {
        when (declaration) {
            is IrModule.Declaration.Function -> {
                // Declares the function in the LLVM module without creating a body for it
                val functionType = IrModule.Declaration.Type.fromDeclaration(declaration)

                // the parent type is a pointer to the parent struct
                val parentType = parentStruct?.let {
                    IrModule.Declaration.Type.Ref(IrModule.Declaration.Type.Struct(it, mapOf()))
                }
                val llvmFunctionType = getType(functionType, parentType)!!

                // prepend the parent struct name to the function name
                val functionName =
                    if (parentStruct == null) declaration.name else parentStruct.name + "_" + declaration.name

                val llvmFunction = LLVMAddFunction(llvmModule, functionName, llvmFunctionType)
                declarationCache[declaration] = llvmFunction to llvmFunctionType
            }
            is IrModule.Declaration.Struct -> {
                val structType = IrModule.Declaration.Type.fromDeclaration(declaration)
                val llvmStructType = getType(structType)!!
                declarationCache[declaration] = null to llvmStructType
                declaration.declarations.forEach { processDeclarationHeader(it, declaration) }
            }
            else -> {
            }
        }
    }

    private fun processDeclaration(
        declaration: IrModule.Declaration,
        parentStruct: IrModule.Declaration.Struct? = null
    ) {
        when (declaration) {
            is IrModule.Declaration.Function -> {
                val (llvmValueRef, llvmTypeRef) = declarationCache[declaration]!!
                LLVMSetFunctionCallConv(llvmValueRef, LLVMCCallConv);
                processFunction(declaration, llvmValueRef!!, llvmTypeRef, parentStruct)
            }
            is IrModule.Declaration.Struct -> {
                declaration.declarations.forEach { processDeclaration(it, declaration) }
            }
        }
    }

    private fun processFunction(
        function: IrModule.Declaration.Function,
        llvmFunction: LLVMValueRef,
        llvmFunctionType: LLVMTypeRef,
        parentStruct: IrModule.Declaration.Struct? = null
    ) {
        // generate llvm blocks for every ir block
        // we do this before generating the instructions because instructions can refer to
        // blocks, like jump and branch
        val blockMap = function.blocks.map { block ->
            block to LLVMAppendBasicBlockInContext(context, llvmFunction, "block${block.id}")
        }.toMap()
        val insMap = mutableMapOf<IrModule.Instruction, LLVMValueRef>()

        function.blocks.forEach { block ->
            LLVMPositionBuilderAtEnd(builder, blockMap[block])

            block.instructions.forEach { ins ->
                val llvmIns = when (ins) {
                    is IrModule.Instruction.IntegerLiteral ->
                        LLVMConstInt(getType(IrModule.Declaration.Type.Primitive.Int32), ins.value, 0)
                    is IrModule.Instruction.BooleanLiteral ->
                        LLVMConstInt(getType(IrModule.Declaration.Type.Primitive.Bool), if (ins.value) 1 else 0, 0)
                    is IrModule.Instruction.Alloca -> {
                        val type = getType(ins.declarationRef)
                        LLVMBuildAlloca(builder, type, ins.name)
                    }
                    is IrModule.Instruction.Load -> {
                        val ptr = insMap[ins.ptr]
                        LLVMBuildLoad(builder, ptr, "")
                    }
                    is IrModule.Instruction.Store -> {
                        val ptr = insMap[ins.ptr]
                        val value = insMap[ins.value]
                        LLVMBuildStore(builder, value, ptr)
                    }
                    is IrModule.Instruction.FloatLiteral -> null
                    is IrModule.Instruction.GetParameter -> {
                        val arg = function.arguments.withIndex().first { it.value.name == ins.name }
                        var index = arg.index
                        // if there's a parent struct, the function's first argument is the struct
                        if (parentStruct != null) {
                            index++
                        }
                        LLVMGetParam(llvmFunction, index)
                    }
                    is IrModule.Instruction.DeclarationReference -> {
                        // todo: support declarations outside module
                        ins.declarationRef.resolve()
                            ?.let { if (it is IrModule.Declaration.Type.Struct) it.declaration else it }
                            ?.let { if (it is IrModule.Declaration.Type.Function) it.declaration else it }
                            ?.let { declarationCache[it]?.first }
                    }
                    is IrModule.Instruction.FunctionCall -> {
                        val func = insMap[ins.function]
                        val args = ins.arguments.map { insMap[it] }
                        val argPtr = PointerPointer<LLVMValueRef>(*args.toTypedArray())
                        LLVMBuildCall(builder, func, argPtr, args.size, "")
                    }
                    is IrModule.Instruction.GetField -> {
                        // get the inner aggregate type
                        val aggregate = insMap[ins.aggregate]
                        val aggregateType = ins.aggregate.getType(function, parentStruct)
                        var innerType = aggregateType
                        while (innerType is IrModule.Declaration.Type.Ref) {
                            innerType = innerType.inner
                        }

                        // get the index of the field within the struct
                        val index = if (innerType is IrModule.Declaration.Type.Struct) {
                            innerType.declaration.fields.withIndex()
                                .firstOrNull { it.value.name == ins.field }?.index ?: -2
                        } else -1

                        if (aggregateType is IrModule.Declaration.Type.Ref) {
                            // build the pointer and load the value
                            val gep = LLVMBuildStructGEP2(builder, getType(innerType!!), aggregate, index, "")
                            LLVMBuildLoad(builder, gep, "")
                        } else {
                            // extract the value directly
                            LLVMBuildExtractValue(builder, aggregate, index, "")
                        }
                    }
                    is IrModule.Instruction.Return -> {
                        val value = insMap[ins.value]
                        LLVMBuildRet(builder, value)
                    }
                    is IrModule.Instruction.Jump -> {
                        LLVMBuildBr(builder, blockMap[ins.block])
                    }
                    is IrModule.Instruction.Branch -> {
                        val cond = insMap[ins.condition]
                        val trueBlock = blockMap[ins.trueBlock]
                        val falseBlock = blockMap[ins.falseBlock]
                        LLVMBuildCondBr(builder, cond, trueBlock, falseBlock)
                    }
                    is IrModule.Instruction.Dereference -> null
                    is IrModule.Instruction.AddressOf -> null
                    is IrModule.Instruction.New -> null
                    is IrModule.Instruction.MethodCall -> {
                        // get the llvm function for this method
                        val method: IrModule.Declaration.Function =
                            ins.getMethod(compiler, irModule, function, parentStruct)!!
                        val func = declarationCache[method]?.first

                        // build the args and append the struct to the beginning
                        val args = ins.arguments.map { insMap[it] }.toMutableList()
                        val aggregate = insMap[ins.aggregate]
                        args.add(0, aggregate)
                        val argPtr = PointerPointer<LLVMValueRef>(*args.toTypedArray())

                        LLVMBuildCall(builder, func, argPtr, args.size, "")
                    }
                    is IrModule.Instruction.BinaryOperator.Add -> null
                    is IrModule.Instruction.BinaryOperator.Sub -> null
                    is IrModule.Instruction.BinaryOperator.Mul -> null
                    is IrModule.Instruction.BinaryOperator.Div -> null
                    IrModule.Instruction.This -> LLVMGetParam(llvmFunction, 0)
                }
                if (llvmIns != null) {
                    insMap[ins] = llvmIns
                }
            }
        }
    }

    fun dump() {
        LLVMDumpModule(llvmModule)
    }

    // useful extension functions

    private fun IrModule.Instruction.getType(
        currentFunction: IrModule.Declaration.Function,
        parent: IrModule.Declaration.Struct? = null
    ) = getType(compiler, irModule, currentFunction, parent)

    private fun IrModule.DeclarationRef.resolve(): IrModule.Declaration? = compiler.resolve(irModule, this)

    private fun IrModule.DeclarationRef.toType(): IrModule.Declaration.Type? =
        compiler.resolve(irModule, this)
            ?.let { if (it is IrModule.Declaration.Type) it else null }
}
